fix Issue 14283 - spurious 'this' is not an lvalue deprecation for auto ref

MartinNowak opened the pull request:
- suppress deprecation message for auto ref parameters as passing
  'this' to auto ref will remain correct when it becomes an rvalue
  and because it's very unlikely that the passing as ref argument
  was intentional

[Issue 14283 – Spurious "this is not an lvalue" deprecation message for "auto ref" arguments](https://issues.dlang.org/show_bug.cgi?id=14283)

MartinNowak commented:
This is clearly a hack, but passing a `this` pointer via `ref` to an `auto ref` parameter is most likely unintentional. Leaving out the deprecation message for `auto ref` introduces a very tiny chance that we break code when switching to rvalue `this` references, but requiring people to add a workaround for perfectly fine and common code would be a worse choice.

WalterBright commented on src/expression.c:
```diff
@@ -3598,6 +3605,12 @@ Expression *ThisExp::toLvalue(Scope *sc, Expression *e)
     return this;
 }
 
+Expression *ThisExp::toLvalue14283(Scope *sc, Expression *e)
```
I see no point in making this a separate function. Just inline it with its comment.

MartinNowak replied:
done

MartinNowak committed:
fix Issue 14283 - spurious 'this' is not an lvalue deprecation for auto ref

- suppress deprecation message for auto ref parameters as passing
  'this' to auto ref will remain correct when it becomes an rvalue
  and because it's very unlikely that the passing as ref argument
  was intentional

WalterBright commented:
Auto-merge toggled on
